### PR TITLE
feat: add coordinated checkpoint prefetch for network filesystem loading

### PR DESCRIPTION
**Pull request by janbernloehr**
Closes #20842

## Motivation

When multiple DP ranks on the same node load the same checkpoint via mmap (e.g. DP-attention), each rank independently page-faults all safetensors files over NFS/Lustre. With N ranks this causes N × checkpoint_size bytes of redundant network I/O.

Note: this is orthogonal to #20332, which addresses the per-rank access pattern problem with TP (striped reads within safetensors). This PR addresses the cross-rank duplication problem with DP (N ranks redundantly reading the same files over NFS). The two optimizations are complementary.

## Modifications

- **`weight_utils.py`**: Added `_PREFETCH_BLOCK_SIZE` module constant, `_prefetch_checkpoint_file()` (reads a file in blocks to warm page cache) and `_prefetch_all_checkpoints()` (distributes files across ranks via `sorted_files[rank::world_size]`, runs in a background daemon thread with configurable thread count). Added `prefetch` and `prefetch_num_threads` parameters to `safetensors_weights_iterator()` and `buffered_multi_thread_safetensors_weights_iterator()`.
- **`server_args.py`**: Added `--weight-loader-prefetch-checkpoints` flag and `--weight-loader-prefetch-num-threads` (default: 4).
- **`loader.py`**: Wired the new flags through to both weight iterator functions.
- **`test_prefetch_checkpoints.py`**: Unit tests verifying byte-level read completeness, correct rank-based file partitioning, and bit-identical weights with/without prefetch.

The prefetch runs in a background daemon thread (inspired by vllm-project/vllm#36012) that pipelines I/O with loading — loading starts immediately while the prefetch thread reads ahead into the OS page cache. Each rank prefetches 1/Nth of the checkpoint shards, reducing total NFS I/O from N*checkpoint to 1*checkpoint.

## Accuracy Tests

No changes to model forward code or kernels. The prefetch only affects the I/O path during weight loading — tensors loaded are identical (verified by unit test `test_weights_match_with_and_without_prefetch`).

## Benchmarking and Profiling

DeepSeek R1-671B FP8 on Lustre, multiple platforms and configurations:

| Platform | Baseline | + Background prefetch | Improvement |
|---|---|---|---|
| DGX B200 (8-GPU, dpa-on) | 1677s (28 min) | **167s (2.8 min)** | **10x** |
| DGX B300 (8-GPU, dpa-on) | 2390s (40 min) | **280s (4.7 min)** | **8.5x** |
| GB200 NVL (2-node, dpa-on) | 263s (4.4 min) | **131s (2.2 min)** | **2x** |
| GB300 NVL (2-node, dpa-on) | 279s (4.6 min) | **152s (2.5 min)** | **1.8x** |
| DGX H100 (2-node TP=16, dpa-off) | 2571s (42.9 min) | **793s (13.2 min)** | **3.2x** |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: test/registered/unit/model_loader/test_prefetch_checkpoints.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+import safetensors.torch
+import torch
+
+from sglang.srt.model_loader.weight_utils import (
+    _prefetch_all_checkpoints,
+    _prefetch_checkpoint_file,
+    safetensors_weights_iterator,
+)
+
```
Comment (Fridge003) [inline test/registered/unit/model_loader/test_prefetch_checkpoints.py:20]: Register this test with `register_cpu_ci(est_time=x, suite="stage-a-test-cpu")`, where x is the estimated running time.

File: test/registered/unit/model_loader/test_prefetch_checkpoints.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+class TestPrefetchReadsAllBytes(unittest.TestCase):
```
Comment (Fridge003) [inline test/registered/unit/model_loader/test_prefetch_checkpoints.py]: This subtest is unnecessary with all the subtests below

File: test/registered/unit/model_loader/test_prefetch_checkpoints.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            ):
+                _prefetch_checkpoint_file(path)
+
+            self.assertEqual(bytes_read, file_size)
+        finally:
+            os.unlink(path)
+
+
+class TestPrefetchDistributedOnlyReadsSubset(unittest.TestCase):
+    """Verify each rank only prefetches its assigned fraction of files."""
```
Comment (Fridge003) [inline test/registered/unit/model_loader/test_prefetch_checkpoints.py]: Also remove this test. We need a multi-gpu test (maybe 4-gpu-h100 test) to protect this feature

File: python/sglang/srt/server_args.py [hunk 4]
```diff
@@ -5546,6 +5506,20 @@ def add_cli_args(parser: argparse.ArgumentParser):
             action="store_true",
             help="Disable mmap while loading weight using safetensors.",
         )
+        parser.add_argument(
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py:6083]: Also update the document in `docs/advanced_features/server_arguments.md`

File: python/sglang/srt/model_loader/weight_utils.py [hunk 5]
```diff
@@ -710,16 +713,121 @@ def np_cache_weights_iterator(
...
+        world_size = torch.distributed.get_world_size()
+    else:
+        rank = 0
+        world_size = 1
+
+    my_files = sorted_files[rank::world_size]
+    total_for_rank = len(my_files)
+
+    if rank == 0:
+        logger.info(
```
Comment (Fridge003) [inline python/sglang/srt/model_loader/weight_utils.py]: Print the log on every rank

File: python/sglang/srt/model_loader/weight_utils.py [hunk 6]
```diff
@@ -710,16 +713,121 @@ def np_cache_weights_iterator(
...
+                )
+
+        await asyncio.gather(*(prefetch_one(p) for p in my_files))
+
+    def _run_prefetch() -> None:
+        start = time.perf_counter()
+        asyncio.run(_prefetch_all())
+        elapsed = time.perf_counter() - start
+        if rank == 0:
+            logger.info(
```
Comment (Fridge003) [inline python/sglang/srt/model_loader/weight_utils.py]: Print this log on every rank

File: python/sglang/srt/model_loader/weight_utils.py [hunk 7]
```diff
@@ -67,6 +67,9 @@
 
 logger = logging.getLogger(__name__)
 
+# Block size for sequential checkpoint prefetch reads (page cache warming).
+_PREFETCH_BLOCK_SIZE = 16 * 1024 * 1024  # 16 MB
```
Comment (Fridge003) [inline python/sglang/srt/model_loader/weight_utils.py]: Can we add an environ that tunes this block size

File: python/sglang/srt/server_args.py [hunk 8]
```diff
@@ -697,6 +697,8 @@ class ServerArgs:
     # For model weight update and weight loading
     custom_weight_loader: Optional[List[str]] = None
     weight_loader_disable_mmap: bool = False
+    weight_loader_prefetch_checkpoints: bool = False
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py:722]: Is there any risk of setting this to True by default

File: python/sglang/srt/model_loader/weight_utils.py [hunk 9]
```diff
@@ -710,16 +713,121 @@ def np_cache_weights_iterator(
...
+
+        async def prefetch_one(path: str) -> None:
+            nonlocal completed, next_log_pct
+            try:
+                async with semaphore:
+                    await asyncio.to_thread(_prefetch_checkpoint_file, path)
+                completed += 1
+                if rank == 0 and total_for_rank > 0 and next_log_pct <= 100:
+                    pct = 100 * completed / total_for_rank
+                    if pct >= next_log_pct:
```
Comment (Fridge003) [inline python/sglang/srt/model_loader/weight_utils.py:787]: Print this on every rank

**Review by Fridge003 — CHANGES_REQUESTED** [review]

File: python/sglang/srt/server_args.py (hunk 8 above)
Comment (janbernloehr) [reply python/sglang/srt/server_args.py:722]: the risk is low but not zero. if e.g. the disk is slow, then prefetch and loader compete for the same I/O bandwidth which could worsen load times. For fast local disks (nvme) or network storage (nfs/lustre) and sufficient host ram this is safe to enable though. I would stay with default false for now - similar to what vllm did.

**Comment by nvpohanh** [comment]
@janbernloehr could you let us know when this PR is ready for another round of review?

**Comment by janbernloehr** [comment]
@nvpohanh Thanks for the ping, I addressed everything. If there is anything else I can do please let me know.

**Comment by nvpohanh** [comment]
@Fridge003 Could you review again? Thanks!

**Review by Fridge003 — APPROVED** [review]
